File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let res = []
    let top = 0
    let bottom = matrix.length // one over
    let left = 0
    let right = matrix[0].length // one over

    while(left < right && top < bottom){
        // add top row
        for(let i = left; i < right; i++){
            res.push(matrix[top][i])
        }
        top++
        // add right col
        for(let i = top; i < bottom; i++){
            res.push(matrix[i][right - 1])
        }
        right--
        if(!(left < right && top < bottom)){
            break
        }
        // add bottom row
        for(let i = right - 1; i >= left; i--){
            res.push(matrix[bottom - 1][i])
        }
        bottom--
        // add left col
        for(let i = bottom - 1; i >= top; i--){
            res.push(matrix[i][left])
        }
        left++
    }
    return res
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])

        top = 0
        bottom = rows - 1
        left = 0
        right = cols - 1

        res = []

        while top <= bottom and left <= right:
            for c in range(left, right +1):
                res.append(matrix[top][c])
            top += 1
            
            for r in range(top, bottom+1):
                res.append(matrix[r][right])
            right -= 1

            if top <= bottom and left <= right:
                for c in range(right,left-1,-1):
                    res.append(matrix[bottom][c])
                bottom -= 1

                for r in range(bottom,top-1,-1):
                    res.append(matrix[r][left])
                left += 1
        
        return res

        